### PR TITLE
fix: support pyarrow input types

### LINKED ISSUE #1625: bug(python): Unsupported data type: <class 'pyarrow.lib.RecordBatch'>
### LanceDB version

0.13.0b1

### What happened?

```
record_batch = pa.RecordBatch.from_pydict(first_batch)
table.add(record_batch)
```
Throws 
```
Traceback (most recent call last):
  File "/Users/lu/Projects/Test/Python/lanceTest/lancedb_test.py", line 37, in <module>
    table.add(record_batch)
  File "/Users/lu/Projects/Test/Python/lanceTest/.venv/lib/python3.10/site-packages/lancedb/table.py", line 1332, in add
    data, _ = _sanitize_data(
  File "/Users/lu/Projects/Test/Python/lanceTest/.venv/lib/python3.10/site-packages/lancedb/table.py", line 125, in _sanitize_data
    raise TypeError(f"Unsupported data type: {type(data)}")
TypeError: Unsupported data type: <class 'pyarrow.lib.RecordBatch'>
```

### Are there known steps to reproduce?

Install the latest lancedb and run the following script will throw the exception.
```
import lancedb
from datasets import load_dataset
import pyarrow as pa
import time
import os
import shutil

print(lancedb.__version__)

dataset_path = "./db/dbpedia_entities_streaming.lance"

if os.path.exists(dataset_path):
    print(f"Removing existing dataset at {dataset_path}")
    shutil.rmtree(dataset_path)

# Define the schema for the LanceDB table
schema = pa.schema([
   pa.field("_id", pa.string()),
   pa.field("title", pa.string()),
   pa.field("text", pa.string()),
   pa.field("openai", pa.list_(pa.float32(), 1536))
])

# Get the first batch
dataset = load_dataset("KShivendu/dbpedia-entities-openai-1M", split="train", streaming=True)

first_batch = next(dataset.iter(batch_size=1000))
record_batch = pa.RecordBatch.from_pydict(first_batch)

# Create LanceDB database
db_path = "/Users/lu/Projects/Test/Python/lanceTest/db"
db = lancedb.connect(db_path)

table_name = "dbpedia_entities_streaming"
table = db.create_table(table_name, schema=schema)

table.add(record_batch)

print("LanceDB dataset created at:", dataset_path)

print("Row count:", table.count_rows())
```
Note that Lance will not throw this exception. The following code works
```
record_batch = pa.RecordBatch.from_pydict(first_batch)

lanceds = lance.write_dataset(
    data_obj=record_batch,
    schema=schema,
    uri=dataset_path,
    mode="create"
)
```

### PR DESCRIPTION
**Pull request by LuQQiu**
fixes #1625 
Support PyArrow.RecordBatch, pa.dataset.Dataset, pa.dataset.Scanner, paRecordBatchReader

### CODE REVIEW

**Comment by LuQQiu** [comment]
TODO the following data types
- pa.RecordBatchReader. <-- created an empty data
- LanceDataset <-- run into segment fault

File: python/python/lancedb/table.py [hunk 1]
```diff
@@ -127,7 +127,26 @@ def _sanitize_data(
...
+        if schema is None:
+            schema = data.schema
+    elif isinstance(data, LanceDataset):
+        data = data.scanner().to_reader()
+        if schema is None:
+            schema = data.schema
+    elif isinstance(data, pa.RecordBatchReader):
+        if schema is None:
+            schema = data.schema
+    elif not isinstance(data, pa.RecordBatchReader):
```
Comment (LuQQiu) [inline python/python/lancedb/table.py:149]: The code is ugly, will solve the bugs first before considering reformat the whole _sanitize_data

Comment (eddyxu) [reply python/python/lancedb/table.py:149]: We have similar code path in lance i believe. Can we reuse them>

**Review by westonpace — APPROVED** [review]
Here's the code path in lance: https://github.com/lancedb/lance/blob/c0e1f158609628f08239520596d90ba3d8c0477f/python/python/lance/dataset.py#L2948

We could, in theory, make the function in `pylance` public utility but I think we want to someday reach a point where `pylance` is an optional dependency for `lancedb` (only used when the user calls `to_lance` and so I'm not sure that would work.

We could maybe add helper functions of some kind to pyarrow (e.g. a `to_reader` for all those types).

In the meantime I think this approach is fine.

**Comment by LuQQiu** [comment]
> Here's the code path in lance: https://github.com/lancedb/lance/blob/c0e1f158609628f08239520596d90ba3d8c0477f/python/python/lance/dataset.py#L2948
> 
> We could, in theory, make the function in `pylance` public utility but I think we want to someday reach a point where `pylance` is an optional dependency for `lancedb` (only used when the user calls `to_lance` and so I'm not sure that would work.
> 
> We could maybe add helper functions of some kind to pyarrow (e.g. a `to_reader` for all those types).
> 
> In the meantime I think this approach is fine.

@westonpace The lancedb input data sanitization involves data change (e.g. add vector embedding column, add schema metadata, change vector column data type, fill in error vector). The original approach is using PyArrow.Table to modify whole data in memory. My current refactor reuses this approach. Let me know if we want to change to ArrowReader and we modify data on the fly (when reading batch records)

**Comment by LuQQiu** [comment]
=========================== short test summary info ===========================
FAILED python\python\tests\test_table.py::test_restore_consistency - assert 2 == 3
 +  where 2 = LanceTable(connection=LanceDBConnection(C:\Users\runneradmin\AppData\Local\Temp\pytest-of-runneradmin\pytest-0\test_restore_consistency0, read_consistency_interval=datetime.timedelta(0)), name="my_table").version
 +  and   3 = LanceTable(connection=LanceDBConnection(C:\Users\runneradmin\AppData\Local\Temp\pytest-of-runneradmin\pytest-0\test_restore_consistency0), name="my_table").version
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
==== 1 failed, 124 passed, 18 skipped, 19 deselected, 3 warnings in 29.48s ====

**Comment by LuQQiu** [comment]
@westonpace PTAL thanks!

**Pull request merged**